lib/pyinotify.py

```diff
@@ -0,0 +1,2346 @@
+        """
+        If the event affects a directory and the auto_add flag of the
+        targetted watch is set to True, a new watch is added on this
+        new directory, with the same attribute values than those of
+        this watch.
+        """
+        if raw_event.mask & IN_ISDIR:
+            watch_ = self._watch_manager.get_watch(raw_event.wd)
+            created_dir = os.path.join(watch_.path, raw_event.name)
+            if watch_.auto_add and not watch_.exclude_filter(created_dir):
+                addw = self._watch_manager.add_watch
+                # The newly monitored directory inherits attributes from its
+                # parent directory.
+                addw_ret = addw(created_dir, watch_.mask,
+                                proc_fun=watch_.proc_fun,
+                                rec=False, auto_add=watch_.auto_add,
+                                exclude_filter=watch_.exclude_filter)
+
+                # Trick to handle mkdir -p /d1/d2/t3 where d1 is watched and
+                # d2 and t3 (directory or file) are created.
+                # Since the directory d2 is new, then everything inside it must
+                # also be new.
+                created_dir_wd = addw_ret.get(created_dir)
+                if ((created_dir_wd is not None) and (created_dir_wd > 0) and
+                    os.path.isdir(created_dir)):
+                    try:
+                        for name in os.listdir(created_dir):
+                            inner = os.path.join(created_dir, name)
+                            if self._watch_manager.get_wd(inner) is not None:
+                                continue
+                            # Generate (simulate) creation events for sub-
+                            # directories and files.
+                            if os.path.isfile(inner):
+                                # symlinks are handled as files.
+                                flags = IN_CREATE
+                            elif os.path.isdir(inner):
+                                flags = IN_CREATE | IN_ISDIR
+                            else:
+                                # This path should not be taken.
+                                continue
+                            rawevent = _RawEvent(created_dir_wd, flags, 0, name)
+                            self._notifier.append_event(rawevent)
+                    except OSError as err:
+                        msg = "process_IN_CREATE, invalid directory: %s"
+                        log.debug(msg % str(err))
+        return self.process_default(raw_event)
+
+    def process_IN_MOVED_FROM(self, raw_event):
+        """
+        Map the cookie with the source path (+ date for cleaning).
+        """
+        watch_ = self._watch_manager.get_watch(raw_event.wd)
+        path_ = watch_.path
+        src_path = os.path.normpath(os.path.join(path_, raw_event.name))
+        self._mv_cookie[raw_event.cookie] = (src_path, datetime.now())
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. The recommended way to create an aware datetime object representing a specific timestamp in UTC is by passing `tzinfo` as an argument to the method.

[Learn more](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)